### ISSUE TITLE
Print or display an old site URL or provide insights along with the "Looks like your URL has changed."

### ISSUE THREAD

**Issue by jamesozzie** [google/site-kit-wp#4247]
## Bug Description

Many users open support topics when they encounter the "_Looks like your URL has changed_" notice, in many cases unsure of such a URL change. This can occur in the following situations:
- When a site has both it's http and https site available, without redirection
- When using a language plugin that may add a /en or other suffix to a homepage
- For site migrations or domain changes

In addition to the error message it might be useful to print the old site URL, in the form of a console error or using a description along the lines of the below:
_..Looks like your URL has changed from ***** to *****._ 

This may help users determine a plugin related issue or identify a https status change. 

[Example support topic](https://wordpress.org/support/topic/looks-like-the-url-of-your-site-has-changed-in-order-to-continue-using-site-kit-2/#post-14957342)

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

Update the current notice to include the old and new URLs:

_Reconnect Site Kit_
_Looks like the URL of your site has changed. In order to continue using Site Kit, you’ll need to reconnect, so that your plugin settings are updated with the new URL._
_- Old URL: https://www.example.com_
_- New URL: https://www.updatedexample.com_

## Implementation Brief

The final implementation changed from the original definition. See https://github.com/google/site-kit-wp/pull/4565

<details>
	<summary>Original IB</summary>
	
* Create `includes/Core/Authentication/Previous_Connected_Proxy_URL.php` which contains the `Previous_Connected_Proxy_URL` class.
	* The class extends the `Connected_Proxy_URL` class.
	* Set the `OPTION` to the `OPTION` of the parent class (`Connected_Proxy_URL::OPTION`), with `_previous` appended.
	* Add the `register` public function which:
		* is hooked to the `update_option_{$option}` action where `$option` is the `Connected_Proxy_URL::OPTION`.
			* The callback function accepts 1 parameter which is the old value for that option and uses the `set` method to set the value of the option to the old value of the parent option.
* Using `includes/Core/Authentication/Authentication.php`,
	* Add a new protected property `previous_connected_proxy_url`.
	* In the constructor, set `previous_connected_proxy_url` to a new instance of the `Previous_Connected_Proxy_URL` class.
	* In the `register` function, call the `register` method of `previous_connected_proxy_url`
	* Update the `get_rest_routes` method to include the above options in the array returned for the `core/user/data/authentication` REST route. The keys can be:
		* `connectedProxyURL` which takes the value of `$this->connected_proxy_url->get()`
		* `previousConnectedProxyURL` which takes the value of `$this->previous_connected_proxy_url->get()`
* Using `assets/js/googlesitekit/datastore/user/authentication.js`,
	* Add the following selectors: `getConnectedProxyURL` and `getPreviousConnectedProxyURL` which retrieves the `connectedProxyURL` and `previousConnectedProxyURL` values similar to how `getDisconnectedReason` is implemented.
* Using `assets/js/components/setup/SetupUsingProxy.js`,
	* Get the `connectedProxyURL` and `previousConnectedProxyURL` values from selectors added above.
	* Display the URLs as per the AC when there is an URL mismatch.

</details>

### Test Coverage
* Add tests for the new selectors.
* Add tests for the new `Previous_Connected_Proxy_URL` class.

## QA Brief

Check that the "_Looks like your URL has changed_" notice is updated to include new and old URLs when a change is detected. This notice is shown in 2 places:
- WP admin (on non-SK screens - e.g. WP dashboard)
- SK splash screen  
Once a URL change is detected, the user will be disconnected immediately and so they'll have to set up Site Kit again to use the plugin

I can provide a simple plugin that when active will simulate a URL change to trigger this. _Changing the URL in WP settings will not trigger this behavior_.

If the "new and old" urls are different, the notice shown should include them. If the URLs happen to be the same but the user is still disconnected from a detected change, the message would look as it does today (without new and old URLs). The latter may sound a bit strange, but is possible if a URL change was detected, causing a disconnect which in turn raises the notice to be shown. In this case, once the notice is triggered, the plugin could be disabled to simulate matching URLs again, the page could then be refreshed to see the message without added URLs.

Finally, disable the simulator plugin and connect Site Kit again to see the notice disappear. Alternatively resetting Site Kit would also clear the notice.

## Changelog entry

* Add information about old and new site URLs to the splash screen when URL mismatch is detected.

**Comment by marrrmarrr** [comment]
@felixarntz LMK if you have any further questions on this one, otherwise let's move it to IB.

**Comment by asvinb** [comment]
@aaemnnosttv @eugene-manuilov  I investigated and looks like we don't store the URLs when there is a mismatch. We only store the reason here:
https://github.com/google/site-kit-wp/blob/develop/includes/Core/Authentication/Authentication.php#L1274-L1275

Not sure if there's a better way of handling this.

**Comment by aaemnnosttv** [comment]
Thanks @asvinb, I think we can simplify the approach a little bit by only keeping track of the last/previous connected proxy URL rather than potentially many previous URLs. We can also decouple the URL itself from the concept of disconnection again by relating it to the connected URL as it may be useful in other situations as well 🙂 

How about we create a `Previous_Connected_Proxy_URL` class to store the old/previous value whenever the `Connected_Proxy_URL` changes. `Previous_Connected_Proxy_URL` could probably even be a subclass of the latter with a few additions, like suffixing the `OPTION` with `_previous`.

We can automate the setting of this option pretty easily as well without coupling it to `Connected_Proxy_URL` by hooking it into the `update_option_{$option}` action where `$option` is the `Connected_Proxy_URL::OPTION`. This action fires whenever this option's value changes and provides the previous value as the first parameter to the callback, where we could simply call `$this->set($value)`.

Then we just need to expose this to the client (likely with a new selector) for use in this notice.

What do you think?

**Comment by asvinb** [comment]
Thanks @aaemnnosttv . Makes sense. Can you take a look at the updated IB?

Thanks!

**Comment by aaemnnosttv** [comment]
Thanks @asvinb – the IB LGTM although we should be sure to add a test case for `Previous_Connected_Proxy_URL` as well mostly for testing the value is set as expected when the connected proxy URL changes.

Let's allow a little more time for this as well with the estimate 👍

**Comment by asvinb** [comment]
Thanks @aaemnnosttv . Updated as per your comments.

**Comment by aaemnnosttv** [comment]
IB ✅

**Comment by ivankruchkoff** [comment]
Not sure how to setup QA here without `QA:eng` label, but we need:
* `DISCONNECTED_REASON_CONNECTED_URL_MISMATCH` - `select( CORE_USER ).getDisconnectedReason()`
* `http://previous.com` - `select( CORE_USER ).getPreviousConnectedProxyURL()`
* `http://current.com` - `select( CORE_USER ).getConnectedProxyURL()`
* `true` - `select( CORE_SITE ).isUsingProxy()`

to get to:

<img width="1234" alt="Screen Shot 2021-12-10 at 3 34 26 PM" src="https://user-images.githubusercontent.com/373931/145655647-40ff51f7-be30-4549-998f-241424efd895.png">

**Comment by eugene-manuilov** [comment]
> Not sure how to setup QA here without `QA:eng` label...

I think we can try it on a multisite instance with subdomains. I have updated QAB.

**Comment by wpdarren** [comment]
### QA Update: ❌ 

@ivankruchkoff @eugene-manuilov have spent quite a bit of time on QA of this one today.   I've worked closely with @jamesozzie because he has a site set up for this scenario in the QAB. I do not see the new version of the message with the old/new URL. The current message is still appearing on the develop branch. 

![image](https://user-images.githubusercontent.com/73545194/146045322-7725b74f-6cdf-485c-83f9-a189969df1c5.png)

We were wondering if the tester plugin works with subdomains/multisites? Any other ideas that might assist us?

**Comment by ivankruchkoff** [comment]
@wpdarren / @jamesozzie can you paste the values for the following console queries?

```
 window.googlesitekit.data.stores['core/user'].selectors.getDisconnectedReason()   
 window.googlesitekit.data.stores['core/site'].selectors.isUsingProxy()
 window.googlesitekit.data.stores['core/user'].selectors.getPreviousConnectedProxyURL()
 window.googlesitekit.data.stores['core/user'].selectors.getConnectedProxyURL()
```

See my comment further up in the thread for what the values need to be:
https://github.com/google/site-kit-wp/issues/4247#issuecomment-991383839

**Comment by aaemnnosttv** [comment]
## QA ❌ 

@ivankruchkoff @wpdarren I took a quick look at this and it seems that a few things are not working properly.

The notice on non-SK screens appears correctly, but there is a problem on the splash screen in that if there are other connected admins on the site, the user will never see the  "Reconnect Site Kit" case because it's checked after the secondary admin one, which shouldn't take precedence here. https://github.com/google/site-kit-wp/blob/02198785380f22f60be4aa713dee7e7b41bda60b/assets/js/components/setup/SetupUsingProxy.js#L186-L197

I found that the updated message showed as expected when the previous connected URL is set. 
![image](https://user-images.githubusercontent.com/1621608/146165588-fca511a4-656e-47ac-a85a-429c7a41853d.png)

Furthermore, it looks like we missed updating both locations where this text is shown in the IB:
- WP notice shown on non-SK screens (not updated)
- SK Splash page (updated, working)

I'm assuming these should both be updated since the AC isn't specific about it, but it's worth noting that the "Reconnect" link in the WP notice goes directly to the proxy so the user would not see the updated notice in this scenario. I'm not sure if we want to make this into a multi-line notice as it is on the splash page so I'll leave this to @felixarntz to advise.

I also noticed a bit of a problem if the user were to navigate right to the SK dashboard after the URL changes. When a non-matching connected proxy URL is detected, the user is disconnected right away but still lands on the SK dashboard. From what I recall, this should not happen because the user would no longer have permissions to view the dashboard at that point and should be redirected to the splash page. **This is out of scope for this issue** but probably needs a new one to address it.

**Comment by felixarntz** [comment]
@aaemnnosttv Good catch, yes both of the issues you're pointing out should be addressed:
1. The WP admin notice where this applies also needs to be updated accordingly.
2. The check for whether the user is a secondary admin shouldn't take precedence over the disconnected reason. I believe switching those conditions here would be sufficient as a fix.

cc @ivankruchkoff I see you already started a follow-up PR to address 2., can you cover 1. in that as well?

**Comment by ivankruchkoff** [comment]
@felixarntz I've put out https://github.com/google/site-kit-wp/pull/4565 to handle both scenarios.

I discussed with @aaemnnosttv for his testing scenario and it was to add a filter on `home_url`, so I wrote a plugin for the test scenario.
```php
<?php

add_filter( 'home_url', function( $url ) {
    if ( false === strpos( $url, 'http:')) {
        return str_replace( 'https:', 'http:', $url );
    } else {
        return str_replace( 'http:', 'https:', $url );
    }
});
```

The issue we could now face is that I update my site kit plugin with the new version which introduces `Previous Connected Proxy URL class`, but has no value set, I then change my URL via the above filter, and get to a scenario where the url has changed, that previous_connected_url is not set (`false`), hence this alternate codepath for the notification
https://github.com/google/site-kit-wp/pull/4565/files#diff-fd8b5c88edabac5e94ebf3751d5ea639f149dec14a1296a74b6890617a92bf2dR1003

Does it make sense to do this, and if it does make sense, we need to reconcile it with the approach on the JS app here:
https://github.com/google/site-kit-wp/blob/681fc336694ab2cc15221c6952e24792c75b17e0/assets/js/components/setup/SetupUsingProxy.js#L286-L293

**Comment by ivankruchkoff** [comment]
Screenshot of the new message in PHP:
<img width="1289" alt="Screen Shot 2021-12-15 at 12 09 19 PM" src="https://user-images.githubusercontent.com/373931/146257842-01a0f167-12c9-48f6-8366-e18e0a0e9c8a.png">

**Comment by felixarntz** [comment]
Thanks @ivankruchkoff! Two comments for the updated admin notice:
* It's a bit hard to parse the URLs when incorporated into the floating text of that notice. Can you please change it to be closer to how it displays in the splash UI as well? As follows:
    * A separate list (`ul`) below the `p` tag, with both URLs prefixed with "Old URL" and "New URL" respectively.
    * Please make sure to correctly capitalize the word "URL".
* Per the above, the copy of the paragraph should then remain exactly the same as it is today (which will also have the benefit of not having to require new translations for this string that was already fine as it is.

While it's still not perfect to have the "Reconnect" link above the two URLs, the URLs are more of a side information anyway for those more "advanced" users who care. Overall, a list will make it easier to read and also provide consistency with the splash UI.

**Comment by ivankruchkoff** [comment]
@felixarntz update for the `<ul>` is out:
<img width="1264" alt="Screen Shot 2021-12-15 at 3 37 34 PM" src="https://user-images.githubusercontent.com/373931/146281767-eaf6946c-1b34-4f72-be54-82d3a3c03a7a.png">

However there is still the pending question from my previous comment:
JS logic:
https://github.com/google/site-kit-wp/blob/681fc336694ab2cc15221c6952e24792c75b17e0/assets/js/components/setup/SetupUsingProxy.js#L286-L293

PHP equivalent:
https://github.com/google/site-kit-wp/pull/4565/files#diff-fd8b5c88edabac5e94ebf3751d5ea639f149dec14a1296a74b6890617a92bf2dR992-R994

Additional scenario in PHP (not in JS), do we need it at all?

> The issue we could now face is that I update my site kit plugin with the new version which introduces Previous Connected Proxy URL class, but has no value set, I then change my URL via the above filter, and get to a scenario where the url has changed, that previous_connected_url is not set (false), 

https://github.com/google/site-kit-wp/pull/4565/files#diff-fd8b5c88edabac5e94ebf3751d5ea639f149dec14a1296a74b6890617a92bf2dR995-R998

**Comment by aaemnnosttv** [comment]
> The issue we could now face is that I update my site kit plugin with the new version which introduces Previous Connected Proxy URL class, but has no value set, I then change my URL via the above filter, and get to a scenario where the url has changed, that previous_connected_url is not set (false),

@ivankruchkoff you're right, this detail was missed during definition. More often than not, I think this will be the case as a manual update of the site's URL would not cause this to happen as we sync the site fields automatically when this happens.

In this case, I think we need to set the previous connected URL from the current in `check_connected_proxy_url` when the disconnect happens as the current saved connected URL is the previous one (connected URL before disconnect).

**Comment by aaemnnosttv** [comment]
Looking into this a bit more, I think the option for the previous connected URL is actually unnecessary as the mismatch is always between the connected proxy URL and the canonical home URL (essentially current `home_url`), both of which are already available in both places.

I've pushed a few updates here in lieu of the timing for the release here, back to @felixarntz  for final review 👍

**Comment by felixarntz** [comment]
@aaemnnosttv Good catch on the previous connected URL - I agree that relying on the canonical home URL makes sense.

**Comment by wpdarren** [comment]
@aaemnnosttv @felixarntz are there any changes to the AC/QAB that I need to be aware of?  There has been a lot of discussion so want to make sure I am covering everything. 😄

**Comment by ivankruchkoff** [comment]
@wpdarren there are two places for the notification to appear, once on a sitekit dashboard url (happens in javascript) and once on a non sitekit page (happens in PHP).

There are screenshots of both in this thread, one message:
![image](https://user-images.githubusercontent.com/1621608/146165588-fca511a4-656e-47ac-a85a-429c7a41853d.png)

![image](https://user-images.githubusercontent.com/373931/146281767-eaf6946c-1b34-4f72-be54-82d3a3c03a7a.png)

Along with the approach in the QAB, you can create a php file plugin/mu-plugin with the following code:
```php
<?php

add_filter( 'home_url', function( $url ) {
    if ( false === strpos( $url, 'http:')) {
        return str_replace( 'https:', 'http:', $url );
    } else {
        return str_replace( 'http:', 'https:', $url );
    }
});
```

**Comment by aaemnnosttv** [comment]
Thanks @ivankruchkoff – I've updated the QAB with a bit more info as well (didn't see you had replied here yet!) 👍

**Comment by wpdarren** [comment]
Thank you @ivankruchkoff @aaemnnosttv :) will pick this up soon.

**Comment by wpdarren** [comment]
### QA Update ⚠️ 

@aaemnnosttv this is looking good other than I am unable to recreate this:

> If the URLs happen to be the same but the user is still disconnected from a detected change, the message would look as it does today (without new and old URLs). The latter may sound a bit strange, but is possible if a URL change was detected, causing a disconnect which in turn raises the notice to be shown. 

I am assuming that the code provided simulates the domain change but I am using a site that has changed so didn't need the plugin.  How can I simulate the test above? 

**Verified:**

Checked that the `Looks like your URL has changed` notice is updated to include new and old URLs when a change is detected. I saw the message on WP dashboard and on SK splash screen. 

<details>
 <summary>Screenshots</summary>
 
![image](https://user-images.githubusercontent.com/73545194/146551887-005a67c2-42fb-4f89-9a57-a79cd8d52a18.png)

![image](https://user-images.githubusercontent.com/73545194/146551904-581fff95-5de2-4363-bc96-5204a0e5d7ba.png)

</details>

**Comment by aaemnnosttv** [comment]
@wpdarren I'm not sure how you did it with the current site you have now but if you could change the URL back to what it was before the disconnect was triggered (without changing via site settings) then that should do it.

Alternatively, you could start with a connected site and use the plugin to trigger the URL change. Once the disconnect is triggered, the plugin can be removed/deactivated and then you should still see the notice but it wouldn't show the new/old URLs any more because they would both be the same.

**Comment by wpdarren** [comment]
### QA Update: ✅ 

**Verified:**

- Checked that the `Looks like your URL has changed` notice is updated to include new and old URLs when a change is detected. I saw the message on WP dashboard and on SK splash screen. 
- Checked that if the URLs happen to be the same but the user is still disconnected from a detected change, the message looks as it does today (without new and old URLs). 

<details>
 <summary>Screenshots</summary>
 
![image](https://user-images.githubusercontent.com/73545194/146551887-005a67c2-42fb-4f89-9a57-a79cd8d52a18.png)

![image](https://user-images.githubusercontent.com/73545194/146551904-581fff95-5de2-4363-bc96-5204a0e5d7ba.png)

</details>
